### ISSUE TITLE
RooHistPdf check

### ISSUE THREAD

**Issue by alpakpinar** [bu-cms/rebsmear#2]
Hey @AndreasAlbert , starting from the code snippet you posted earlier, I developed a small test script here:

https://github.com/alpakpinar/rebsmear/blob/2021-01-27_rootest/rootest.py

Here, I follow the workflow of the main code: Define a derived variable, re-evaluate it as a RooRealVar and feed it into RooDataHist and RooHistPdf. You can see what the code outputs at the bottom of the script. 

I'm not sure if that's how the minimization software acquires the values for each iteration, so the comparison might not be entirely valid. But I see that the values returned from the pdf seem consistent. I'm also attaching a plot of the PDF for reference. I believe getValV() returns the un-normalized (but bin-width divided) values, seems consistent with the docs in [1].

Let me know what you think, thanks!

[1]: https://root.cern.ch/doc/master/classRooAbsPdf.html#a9314c1474257bd73409c6a9353e70102

**Comment by alpakpinar** [comment]
![test](https://user-images.githubusercontent.com/45109652/106057243-b06dbf80-60bd-11eb-9ed6-aa4711016c97.png)
For reference, this is what the PDF looks like

**Comment by alpakpinar** [comment]
I'm actually working on writing a small minimizer on top of the existing script, hoping to test this better. I have some problems for now but I can share the results when I have the minimizer working properly.

**Comment by AndreasAlbert** [comment]
Thanks for the demonstration. I don't think you are testing what we are trying to do though. What you are doing here is:

1. loop over px, py values
2. define an entirely new pt variable every iteration
3. define an entirely new PDF at every iteration and check what its value is

so effectively what you are testing is that RooHistPDF can correctly evaluate a histogram in general (good, but not really what we were curious about). What we *really* want  to demonstrate is this:

1. define px and py at some starting value
2. define the PDF evaluated at pt
3. critical: show that the PDF value changes as we change the underlying variables.

basically this:
```python
(...)
pdf = RooHistPdf(...)
pdf.evaluate() # value we start out at
px.setVal(5)
pdf.evaluate() # new, changed value
py.setVal(-3)
pdf.evaluate() # changed again
```

Because after all, the code we are talking here is run once to build the workspace, but we need to make sure that the propagation works when we later run the minimization.

**Comment by alpakpinar** [comment]
Hey @AndreasAlbert, thank you for this. Based on your comments, I re-adjusted the test in this script:

https://github.com/alpakpinar/rebsmear/blob/2021-01-27_rootest/rootest.py

(pdf.evaluate() doesn't work for me, I think it's a protected member function according to [1], so I'm going with pdf.getValV() )

So in this new test, by changing the primitive value px, I see the RooFormulaVar px^2 adjusts accordingly. But the RooRealVar we define for px^2 stays constant at the earlier value, hence the PDF value doesn't change. Intuitively this seems to make more sense to me, but I'm confused because the prior does not stay constant at our main code (pdf defined in pretty much the same way). So maybe we should wait for ROOT team's response on the forum as well.

[1]: https://root.cern.ch/doc/master/classRooHistPdf.html
